pyqtorch/core/operation.py

```diff
@@ -262,12 +301,152 @@ def CNOT(state: torch.Tensor, qubits: ArrayLike, N_qubits: int) -> torch.Tensor:
+    return ControlledOperationGate(state, qubits, N_qubits, operation_matrix)
+
+
+def S(state: torch.Tensor, qubits: ArrayLike, N_qubits: int) -> torch.Tensor:
+    """S single-qubit gate
+
+    Args:
+        state (torch.Tensor): the input quantum state, of shape `(N_0, N_1,..., N_N, batch_size)`
+        qubits (ArrayLike): list of qubit indices where the gate will operate
+        N_qubits (int): the number of qubits in the system
+
+    Returns:
+        torch.Tensor: the resulting state after applying the gate
+    """
+
+    if ops_cache.enabled:
+        store_operation("S", qubits)
+
+    dev = state.device
+    mat = SMAT.to(dev)
+    return _apply_gate(state, mat, qubits, N_qubits)
+
+
+def T(state: torch.Tensor, qubits: ArrayLike, N_qubits: int) -> torch.Tensor:
+    """T single-qubit gate
+
+    Args:
+        state (torch.Tensor): the input quantum state, of shape `(N_0, N_1,..., N_N, batch_size)`
+        qubits (ArrayLike): list of qubit indices where the gate will operate
+        N_qubits (int): the number of qubits in the system
+
+    Returns:
+        torch.Tensor: the resulting state after applying the gate
+    """
+
+    if ops_cache.enabled:
+        store_operation("T", qubits)
+
+    dev = state.device
+    mat = TMAT.to(dev)
+    return _apply_gate(state, mat, qubits, N_qubits)
+
+
+def SWAP(state: torch.Tensor, qubits: ArrayLike, N_qubits: int) -> torch.Tensor:
+    """SWAP 2-qubit gate
+
+    Args:
+        state (torch.Tensor): the input quantum state, of shape `(N_0, N_1,..., N_N, batch_size)`
+        qubits (ArrayLike): list of qubit indices where the gate will operate
+        N_qubits (int): the number of qubits in the system
+
+    Returns:
+        torch.Tensor: the resulting state after applying the gate
+    """
+
+    if ops_cache.enabled:
+        store_operation("SWAP", qubits)
+
+    dev = state.device
+    mat = SWAPMAT.to(dev)
+    return _apply_gate(state, mat, qubits, N_qubits)
+
+
+def CPHASE(
+    theta: torch.Tensor, state: torch.Tensor, qubits: ArrayLike, N_qubits: int
+) -> torch.Tensor:
+    """Parametrized 2-qubit CPHASE gate
+
+    Args:
+        theta (torch.Tensor): 1D-tensor holding the values of the parameter
+        state (torch.Tensor): the input quantum state, of shape `(N_0, N_1,..., N_N, batch_size)`
+        qubits (ArrayLike): list of qubit indices where the gate will operate
+        N_qubits (int): the number of qubits in the system
+
+    Returns:
+        torch.Tensor: the resulting state after applying the gate
+    """
+    if ops_cache.enabled:
+        store_operation("CPHASE", qubits, param=theta)
 
     dev = state.device
-    mat = torch.tensor(
-        [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0]], dtype=torch.cdouble
+    mat: torch.Tensor = torch.tensor(
+        [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, torch.exp(torch.tensor(1j * theta))]], dtype=torch.cdouble
```

Comment:
Looks good this matrix.

Reply:
Did you try it on a full pipeline @dominikandreasseitz ? Usually pytorch doesn't like to declare matrices that way it can cause issues with backpropagation

Reply:
good point, I will add a test for that

Reply:
@Slimane33, what do you mean by "issues with backpropagation"? It looks to me like a very basic definition that should not cause any problem to the autodiff engine.

Reply:
It was the way I tried to implement at first the other gates but it returns errors because somehow when you do `torch.tensor([torch.exp(v)])` with a variable `v` you break the computational graph. Maybe it has been corrected in newer versions of pytorch.
The correct way to do it would be `mat: torch.Tensor = torch.exp(1j *theta *torch.tensor(
        [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 1]]`